src/psyclone/psyir/nodes/omp_task_directive.py

```diff
@@ -0,0 +1,2136 @@
+            return isinstance(child, OMPDependClause)
+        return False
+
+    @property
+    def input_depend_clause(self):
+        """
+        :returns: the OMPDependClause child of this node corresponding to \
+                  input dependencies.
+        :rtype: :py:class:`psyclones.psyir.nodes.OMPDependClause`
+        """
+        return self.children[4]
+
+    @property
+    def output_depend_clause(self):
+        """
+        :returns: the OMPDependClause child of this node corresponding to \
+                  output dependencies.
+        :rtype: :py:class:`psyclones.psyir.nodes.OMPDependClause`
+        """
+        return self.children[5]
+
+    def begin_string(self):
+        """Returns the beginning statement of this directive, i.e.
+        "omp task ...". The visitor is responsible for adding the
+        correct directive beginning (e.g. "!$").
+
+        :returns: the beginning statement for this directive.
+        :rtype: str
+
+        """
+        # Generate the string containing the required clauses
+        return "omp task"
+
+    def end_string(self):
+        """Returns the end (or closing) statement of this directive, i.e.
+        "omp end task". The visitor is responsible for adding the
+        correct directive beginning (e.g. "!$").
+
+        :returns: the end statement for this directive.
+        :rtype: str
+
+        """
+        # pylint: disable=no-self-use
+        return "omp end task"
+
+    def validate_global_constraints(self):
+        """
+        Perform validation checks that can only be done at code-generation
+        time.
+
+        :raises GenerationError: if this OMPTaskDirective is not \
+                                 enclosed within an OpenMP serial region.
+        """
+        # It is only at the point of code generation that we can check for
+        # correctness (given that we don't mandate the order that a user
+        # can apply transformations to the code). A taskloop
+        # directive, we must have an OMPSerialDirective as an
+        # ancestor back up the tree.
+        if not self.ancestor(OMPSingleDirective):
+            raise GenerationError(
+                "OMPTaskDirective must be inside an OMP Single region "
+                "but could not find an ancestor node."
+            )
+
+        # If there is a nowait clause on the ancestor OMPSingleDirective,
+        # it is possible that we could make other tasks we expect to be
+        # dependent that wouldn't be, as dependencies are only counted
+        # in OpenMP if spawned by the same thread.
+        if self.ancestor(OMPSingleDirective).nowait:
+            raise GenerationError(
+                "OMPTaskDirective found inside an OMP Single region "
+                "with nowait attached. This means we can't guarantee "
+                "correctness with other potential Single regions so is "
+                "forbidden with PSyclone.")
+
+
+class DynamicOMPTaskDirective(OMPTaskDirective):
```

Comment:
This is a big source file. Please could you split this class into a separate file and split the associated tests too (ensuring that each test file gives 100% coverage of its associated source file).

Reply:
Split these now, there's 2 coverage lines missing, one I'm a bit stuck with so I'll comment below.